liquidhaskell-boot/src-ghc/Liquid/GHC/API/Extra.hs

```diff
@@ -170,45 +169,57 @@ apiCommentsParsedSource ps =
 
--- | Adds NOINLINE pragmas to all local bindings in the module.
+-- | Adds NOINLINE pragmas to all bindings in the module.
 --
 -- This prevents the simple optimizer from inlining such bindings which might
 -- have specs that would otherwise be left dangling.
 --
 -- https://gitlab.haskell.org/ghc/ghc/-/issues/24386
 --
-addNoInlinePragmasToLocalBinds :: ParsedModule -> ParsedModule
-addNoInlinePragmasToLocalBinds ps =
-    ps { pm_parsed_source = go (pm_parsed_source ps) }
+addNoInlinePragmasToBinds :: TcGblEnv -> TcGblEnv
+addNoInlinePragmasToBinds tcg = tcg{ tcg_binds = go (tcg_binds tcg) }
   where
     go :: forall a. Data a => a -> a
-    go = gmapT ((id `extT` addNoInlinePragmas) . go)
-
-    addNoInlinePragmas :: HsValBinds GhcPs -> HsValBinds GhcPs
-    addNoInlinePragmas = \case
-      ValBinds x binds sigs ->
-          ValBinds x binds (newSigs ++ dropInlinePragmas sigs)
-        where
-          dropInlinePragmas = filter (not . isInlineSig . unLoc)
-
-          isInlineSig InlineSig{} = True
-          isInlineSig _ = False
-
-          newSigs = concatMap (traverse noInlinePragmaForBind) $ bagToList binds
-
-          noInlinePragmaForBind :: HsBindLR GhcPs GhcPs -> [Sig GhcPs]
-          noInlinePragmaForBind bndr =
-              [ InlineSig
-                  []
-                  (noLocA b)
-                  defaultInlinePragma
-                    { inl_inline = NoInline $ SourceText $ occNameFS $ rdrNameOcc b
-                    , inl_act = NeverActive
-                    }
-              | b <- collectHsBindBinders CollNoDictBinders bndr
-              ]
-
-      bnds -> bnds
+    go = gmapT $ go `extT` markHsBind
+
+    -- Mark all user-originating `Id` binders as `NOINLINE`.
+    markHsBind :: HsBind GhcTc -> HsBind GhcTc
+    markHsBind = \case
+        bind@VarBind{ var_id = var, var_rhs = rhs } -> bind{ var_id = markId var, var_rhs = go rhs }
+        bind@FunBind{ fun_id = var, fun_matches = matches } -> bind{ fun_id = markId <$> var, fun_matches = go matches }
+        bind@PatBind{ pat_lhs = lhs, pat_rhs = rhs } -> bind{ pat_lhs = markPat <$> lhs, pat_rhs = go rhs }
+        PatSynBind{} -> error "markNoInline: unexpected PatSynBind, should have been eliminated by the typechecker"
+        XHsBindsLR absBinds -> XHsBindsLR (markAbsBinds absBinds)
+
+    markPat :: Pat GhcTc -> Pat GhcTc
+    markPat = \case
+        VarPat ext var -> VarPat ext (markId <$> var)
+        pat -> gmapT (id `extT` markPat) pat
+
+    markId :: Id -> Id
+    markId var = var `setInlinePragma` neverInlinePragma
+
+    -- The AbsBinds come from the GHC typechecker to handle polymorphism,
+    -- overloading, and recursion, so those don't correspond directly to
+    -- user-written `Id`s. In fact, marking them as `NOINLINE` results in Core
+    -- that LH can't process.
+    markAbsBinds :: AbsBinds -> AbsBinds
+    markAbsBinds absBinds0 =
+        absBinds0
+          { abs_binds = fmap (fmap skipFirstHsBind) $ abs_binds absBinds0
```

Comment:
Actually, all tests continue to pass if I replace `skipFirstHsBind` with `go` here. So it is only the bit of `markABE` that makes a difference to LH.

Reply:
I edited the motivation for `markAbsBinds` in 25c218a